ystia/yorc
# Feature/gh 220 refactors slurm jobs

# Pull Request description
Refactor Slurm Jobs
As: Yorc dev team
I want to: Refactor how Slurm Jobs are modeled
So that: It could be more user-friendly

AC1: Distinction between srun and sbatch jobs does not make sens anymore and make it more complex
AC2: We should be able to run a job only using command line args without providing an artifact
AC3: Ideally if singularity jobs could be factorized with other jobs TOSCA desc it would be nice. Maybe only the implementation type should be enough to differentiate job types
AC4: Simplify the implementation in YORC
AC4: Retrieve real Slurm job status once runned

## Description of the change
All ACs OK except AC3: Singularity Job still exists with specific properties related to Singularity execution
### What I did
- Job are runned asynchronously so only Slurm sbatch is supported.
- A command can be executed as a job (no need to provide any artifact) with **command** and **args** properties
- Environment variables can be explicitly added with **env_vars** property.
- Add **working_directory** Tosca prop to let users define the working directory for Slurm sbatch execution and check outputs, stderr, stdout relative to this directory
- **keep_job_remote_artifacts** Slurm infrastructure property allows not to remove artifacts when a job is done.
- Slurm scontrol command is used for monitoring job (allows to get job status, stdout/stderr)

### How to verify it

- Create yorc orchestrator
- Create Slurm location
- Add Job and SinglarityJob on-demand resources in Slurm location
- Add https://github.com/ystia/yorc-a4c-plugin/tree/feature/gh-220-refactors-slurm-jobs/tosca-samples to 
  catalog
- Check and deploy the following topologies:

 **For Slurm job**
 - command
 - mpi
 - mpi2
 - failure

**For Singularity Slurm job:**

- hello
- lolcow
- bash_loop
- compute_pi
- mnist

### Description for the changelog
* Refactor Slurm jobs ([GH-220](https://github.com/ystia/yorc/issues/220))

## Applicable Issues
fixes #220
